## Project: UNDP ACCELERATOR LABS

In [143]:
# Importing Packages
import numpy as np
import pandas as pd
import geopandas as gpd
import seaborn as sns
import matplotlib.pyplot as plt
import translators as ts

%matplotlib inline

pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)
pd.options.mode.chained_assignment = None

In [124]:
# If you run into issues installing geopandas, run the following script in your terminal
# pip install pipwin
# pipwin install gdal
# pipwin install fiona
# pip install geopandas

In [5]:
# Loading data
data = pd.read_excel("Viz4SocialGood_submissionV1.xlsx", sheet_name = "Energy_sol_data")

In [110]:
energy_data = data.copy()

In [111]:
energy_data.head()

contribution_date  \
0  Sun May 30 2021 13:48:32 GMT+0000 (Coordinated...   
1  Mon Jun 28 2021 13:56:30 GMT+0000 (Coordinated...   
2  Mon Jun 28 2021 13:32:36 GMT+0000 (Coordinated...   
3  Thu Nov 12 2020 06:03:43 GMT+0000 (Coordinated...   
4  Wed Jul 21 2021 10:12:39 GMT+0000 (Coordinated...   

                   new_date    id          Energy source  \
0  Sun May 30 2021 13:48:32  3100                Thermal   
1  Mon Jun 28 2021 13:56:30  3177                   Wind   
2  Mon Jun 28 2021 13:32:36  3176                  Hydro   
3  Thu Nov 12 2020 06:03:43  2261      Renewable general   
4  Wed Jul 21 2021 10:12:39  3236  Household application   

  Clean cooking application (yes)                                  title  \
0                             NaN                               Eco Char   
1                             NaN                   Wind power generator   
2                             NaN                       Hybrid Generator   
3                             NaN  Electric Power Microgeneration System   
4                               x                           Solar Cooker   

           Mapper Contributor anonymized  \
0  AccLab Algeria          Contributor 1   
1   AccLab Angola          Contributor 2   
2   AccLab Angola          Contributor 3   
3   AccLab Angola          Contributor 4   
4   AccLab Angola          Contributor 5   

  What is the purpose of the solution? (brief problem & solution description)  \
0  We upcycle wastes such as food organic wastes ...                            
1  The system to obtain electrical energy through...                            
2  The production of fuel from water is done thro...                            
3  The electric power microgeneration system is d...                            
4   Solar Cooker\n\n\n\t\t\t\n\n\n\t\t\t\n\n\n\t\t\t                            

  Please insert a link to the solution  \
0                                  NaN   
1                                  NaN   
2                                  NaN   
3                                  NaN   
4                                  NaN   

  This solution is Do it Yourself / open source  \
0                                           NaN   
1                                           NaN   
2                                           NaN   
3                                           NaN   
4                                           NaN   

  This solution is protected by Intellectual Property  \
0                                                NaN    
1                                                NaN    
2                                                NaN    
3                                                NaN    
4                                                NaN    

  The solution holder is able to train others (including end-users) in using or replicating the solution  \
0                                                NaN                                                       
1                                                NaN                                                       
2                                                NaN                                                       
3                                                NaN                                                       
4                                                NaN                                                       

  What is the unit cost of this Solution along with any additional cost for maintenance and training?  \
0                                                NaN                                                    
1                                                NaN                                                    
2                                                NaN                                                    
3                                                NaN                                                    
4                                                Na

In [112]:
energy_data['new_date'] = pd.to_datetime(energy_data['new_date'])
energy_data.sample(5)

contribution_date            new_date  \
104  Fri Sep 18 2020 14:41:07 GMT+0000 (Coordinated... 2020-09-18 14:41:07   
339  Thu Jun 09 2022 09:16:55 GMT+0000 (Coordinated... 2022-06-09 09:16:55   
245  Fri May 27 2022 04:59:50 GMT+0000 (Coordinated... 2022-05-27 04:59:50   
103  Thu Sep 17 2020 16:10:28 GMT+0000 (Coordinated... 2020-09-17 16:10:28   
256  Wed May 18 2022 08:15:03 GMT+0000 (Coordinated... 2022-05-18 08:15:03   

       id          Energy source Clean cooking application (yes)  \
104   990                Thermal                             NaN   
339  5067  Household application                               x   
245  4900                  Solar                             NaN   
103   776                  Solar                             NaN   
256  4790                Thermal                               x   

                                  title                     Mapper  \
104  Bioethanol from agricultural waste             AccLab Ecuador   
339     Energy-Efficient Portable Stove  Honey Bee Network (India)   
245                     Paneles Solares              AccLab Panama   
103    Solar Power River Transportation             AccLab Ecuador   
256            MUNGUAKONKWA TAKA HUBERT              AccLab Rwanda   

    Contributor anonymized  \
104        Contributor 105   
339        Contributor 340   
245        Contributor 246   
103        Contributor 104   
256        Contributor 257   

    What is the purpose of the solution? (brief problem & solution description)  \
104  Finalist in Post-covid hackathon organized by ...                            
339  The innovative stove is a double-chambered eff...                            
245          Generacion o produccion energetica limpia                            
103  Kara Solar is a solar-powered river transporta...                            
256  Problem: Waste management issues and access to...                            

    Please insert a link to the solution  \
104                                  NaN   
339                                  NaN   
245                                  NaN   
103                                  NaN   
256                                  NaN   

    This solution is Do it Yourself / open source  \
104                                           NaN   
339                                           NaN   
245                                           NaN   
103                                           NaN   
256                                           NaN   

    This solution is protected by Intellectual Property  \
104                                                NaN    
339                                                NaN    
245                                                NaN    
103                                                NaN    
256                                                  x    

    The solution holder is able to train others (including end-users) in using or replicating the solution  \
104                                                NaN                                                       
339                                                NaN                                                       
245                                                NaN                                                       
103                                                NaN                                                       
256                                                  x                                                       

    What is the unit cost of this Solution along with any additional cost for maintenance and training?  \
104                                                NaN                                                    
339                                                NaN                                                    
245                                                NaN                                                    
103                           

In [113]:
energy_data.columns

Index(['contribution_date', 'new_date', 'id', 'Energy source',
       'Clean cooking application (yes)', 'title', 'Mapper',
       'Contributor anonymized',
       'What is the purpose of the solution? (brief problem & solution description)',
       'Please insert a link to the solution',
       'This solution is Do it Yourself / open source',
       'This solution is protected by Intellectual Property',
       'The solution holder is able to train others (including end-users) in using or replicating the solution',
       'What is the unit cost of this Solution along with any additional cost for maintenance and training?',
       'This solution is a Prototype', 'This solution is a Product',
       'If this solution is a product, is it available in market?',
       'If this solution is a product, does advance order has to be given?',
       'What is the Technological Readiness Level (TRL) of this solution?',
       'How much has this solution already been diffused? Is there potential fe

In [114]:
energy_data.columns
energy_data.drop('contribution_date', inplace=True, axis=1)

In [115]:
# Here, column names have been renamed to shorter but descriptive forms.
# To save time, the column names were changed to title case so we don't have to rename them in our charts.

energy_data.rename(columns = {'new_date' : 'Date', 
        'id': 'ID',
        'Energy source': 'Energy Source',
       'Clean cooking application (yes)' : 'Clean Cooking',
       'title': 'Project Title',
       'Contributor anonymized' : 'Contributor',
       'What is the purpose of the solution? (brief problem & solution description)' : 'Purpose of Solution',
       'Please insert a link to the solution' : 'Link',
       'This solution is Do it Yourself / open source' : 'Open Source',
       'This solution is protected by Intellectual Property' : 'IP Protected',
       'The solution holder is able to train others (including end-users) in using or replicating the solution' : 'Can Train Others',
       'What is the unit cost of this Solution along with any additional cost for maintenance and training?' : 'Unit Cost in USD',
       'This solution is a Prototype' : 'Prototype', 
        'This solution is a Product' : 'Product',
       'If this solution is a product, is it available in market?' : 'Market Ready',
       'If this solution is a product, does advance order has to be given?' : 'Requires Advanced Order',
       'What is the Technological Readiness Level (TRL) of this solution?' : 'Tech Readiness Level',
       'How much has this solution already been diffused? Is there potential feedback from end-users available?' : 'User Feedback',
       'Please upload a link of end-user feedback' : 'Feedback Link',
       'Are there any efficiency benchmarks for this solution (eg. how much energy does it save; how much cheaper does it produce energy than current market rates/ current household expenditure / cost per kW h)?' : 'Efficiency Benchmark',
       'Are there any other potential bottlenecks affecting cross-border or in country diffusion of this solution?': 'Potential Bottlenecks',
       'What Sustainable Development Goal is this Solution addressing? Tag 1' : 'SDG Solution1',
       'What Sustainable Development Goal is this Solution addressing? Tag 2' : 'SDG Solution2',
       'What Sustainable Development Goal is this Solution addressing? Tag 3' : 'SDG Solution3',
       'What Sustainable Development Goal is this Solution addressing? Tag 4' : 'SDG Solution4',
       'What Sustainable Development Goal is this Solution addressing? Tag 5' : 'SDG Solution5',
       'What thematic tags apply to this solution? Tag 1' : 'Thematic Tag1',
       'What thematic tags apply to this solution? Tag 2' : 'Thematic Tag2',
       'What thematic tags apply to this solution? Tag 3' : 'Thematic Tag3',
       'What thematic tags apply to this solution? Tag 4' : 'Thematic Tag4',
       'What thematic tags apply to this solution? Tag 5' : 'Thematic Tag5', 
        'Latitude ' : 'Lat',
       'Longitude ' : 'Long'}, inplace = True)

In [116]:
energy_data.head()

Date    ID          Energy Source Clean Cooking  \
0 2021-05-30 13:48:32  3100                Thermal           NaN   
1 2021-06-28 13:56:30  3177                   Wind           NaN   
2 2021-06-28 13:32:36  3176                  Hydro           NaN   
3 2020-11-12 06:03:43  2261      Renewable general           NaN   
4 2021-07-21 10:12:39  3236  Household application             x   

                           Project Title          Mapper    Contributor  \
0                               Eco Char  AccLab Algeria  Contributor 1   
1                   Wind power generator   AccLab Angola  Contributor 2   
2                       Hybrid Generator   AccLab Angola  Contributor 3   
3  Electric Power Microgeneration System   AccLab Angola  Contributor 4   
4                           Solar Cooker   AccLab Angola  Contributor 5   

                                 Purpose of Solution Link Open Source  \
0  We upcycle wastes such as food organic wastes ...  NaN         NaN   
1  The system to obtain electrical energy through...  NaN         NaN   
2  The production of fuel from water is done thro...  NaN         NaN   
3  The electric power microgeneration system is d...  NaN         NaN   
4   Solar Cooker\n\n\n\t\t\t\n\n\n\t\t\t\n\n\n\t\t\t  NaN         NaN   

  IP Protected Can Train Others Unit Cost in USD Prototype Product  \
0          NaN              NaN              NaN       NaN     NaN   
1          NaN              NaN              NaN       NaN     NaN   
2          NaN              NaN              NaN       NaN     NaN   
3          NaN              NaN              NaN       NaN     NaN   
4          NaN              NaN              NaN       NaN     NaN   

  Market Ready Requires Advanced Order Tech Readiness Level User Feedback  \
0          NaN                     NaN                  NaN           NaN   
1          NaN                     NaN                  NaN           NaN   
2          NaN                     NaN                  NaN           NaN   
3          NaN                     NaN                  NaN           NaN   
4          NaN                     NaN                  NaN           NaN   

  Feedback Link Efficiency Benchmark Potential Bottlenecks  SDG Solution1  \
0           NaN                  NaN                   NaN            1.0   
1           NaN                  NaN                   NaN            1.0   
2           NaN                  NaN                   NaN           10.0   
3           NaN                  NaN                   NaN            1.0   
4           NaN                  NaN                   NaN            7.0   

   SDG Solution2  SDG Solution3  SDG Solution4  SDG Solution5  \
0            6.0           13.0            NaN            NaN   
1            7.0           13.0            NaN            NaN   
2           12.0           13.0            NaN            NaN   
3            7.0            9.0           10.0           11.0   
4           12.0           13.0            NaN            NaN   

           Thematic Tag1          Thematic Tag2          Thematic Tag3  \
0  food waste management                  youth                 coffee   
1     alternative energy           clean energy         climate change   
2                    NaN     alternative energy            electricity   
3           clean energy  community empowerment  rural electrification   
4             innovation     polution reduction           solar energy   

           Thematic Tag4 Thematic Tag5         Lat        Long        image_1  \
0                 olives           NaN  36.7607349   3.0261856            NaN   
1  community empowerment           NaN  -12.833226  13.6010742   file: img-10   
2     polution reduction           NaN  -8.5158356  16.0817871  file: img-121   
3         social justice       welfare  -12.776115   15.687688  file: img-230   
4                    NaN           NaN   -12.86536  13.8098145  file: img-305   

  image_2 image_3 image_4 image_5 image_6 image_

In [117]:
energy_data['Unit Cost in USD'].value_counts()

Undefined                                                                                                                                                                                                                                                                                         2
Unsure.                                                                                                                                                                                                                                                                                           2
20 USD                                                                                                                                                                                                                                                                                            2
1200 BDT per unit                                                                                                           

In [118]:
# Load statistics of data
print(energy_data.shape)
energy_data.info()

(359, 43)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 359 entries, 0 to 358
Data columns (total 43 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   Date                     357 non-null    datetime64[ns]
 1   ID                       359 non-null    int64         
 2   Energy Source            355 non-null    object        
 3   Clean Cooking            82 non-null     object        
 4   Project Title            359 non-null    object        
 5   Mapper                   359 non-null    object        
 6   Contributor              359 non-null    object        
 7   Purpose of Solution      347 non-null    object        
 8   Link                     126 non-null    object        
 9   Open Source              100 non-null    object        
 10  IP Protected             41 non-null     object        
 11  Can Train Others         127 non-null    object        
 12  Unit Cost in USD         7

### About the data
The data consists of 359 entries for 44 categories.

### Data quality issues
- There are lots of Null values with few incorrect data types
- Invalid data in the lat and long columns [xxxx, xxxxx]
- Non-descriptive values 'x' in columns: ['Clean Cooking', 'Open Source', 'IP Protected', 'Can Train Others', 'Prototype', 'Product', 'Market Ready', 'Requires Advanced Order']
- Columns are in wrong datatypes [ID, ]
- The coulmn 'Unit Cost in USD' consists of a lot of currencies, numbers and texts.
- Some textx are not in English

### Tidiness Issues
- Some columns are opposite of the other, thus should form one column. Affected columns include 
[Prototype, Product	], [Solution1	SDG Solution2	SDG Solution3	SDG Solution4	SDG Solution5], [Thematic Tag1	Thematic Tag2	Thematic Tag3	Thematic Tag4	Thematic Tag5]

In [119]:
energy_data.sample(3)

Date    ID          Energy Source Clean Cooking  \
5  2021-07-22 06:51:22  3240                Thermal           NaN   
61 2022-06-10 14:06:42  5159                Thermal           NaN   
27 2022-05-16 19:35:47  4771  Household application             x   

                                Project Title               Mapper  \
5     Biogas production through a Biodigester        AccLab Angola   
61                             Pompe eolienne  AccLab Burkina Faso   
27  Ecological kitchen: SOLAR INTI Foundation     AccLab Argentina   

       Contributor                                Purpose of Solution  \
5    Contributor 6  Biogas is produced from cattle manure and orga...   
61  Contributor 62  La solution permet de pomper l'eau en utilisan...   
27  Contributor 28  This solution seeks to address the scarcity of...   

                                   Link Open Source IP Protected  \
5                                   NaN         NaN          NaN   
61                                  NaN         NaN            x   
27  https://www.facebook.com/SolarInti/         NaN          NaN   

   Can Train Others Unit Cost in USD Prototype Product Market Ready  \
5               NaN              NaN       NaN     NaN          NaN   
61                x        1450 $ US       NaN       x          NaN   
27              NaN              NaN       NaN     NaN          NaN   

   Requires Advanced Order                               Tech Readiness Level  \
5                      NaN                                                NaN   
61                       x  9. Systeme reel prouve a travers des operation...   
27                     NaN                                                NaN   

                                        User Feedback Feedback Link  \
5                                                 NaN           NaN   
61  Presentation de prototype lors des forums de l...           NaN   
27                                                NaN           NaN   

                                 Efficiency Benchmark Potential Bottlenecks  \
5                                                 NaN                   NaN   
61  Pas de comparaison pour le moment. L"efficacit...                   NaN   
27                                                NaN                   NaN   

    SDG Solution1  SDG Solution2  SDG Solution3  SDG Solution4  SDG Solution5  \
5             7.0           13.0            NaN            NaN            NaN   
61            7.0            8.0           12.0           13.0            NaN   
27            7.0           10.0           11.0            NaN            NaN   

            Thematic Tag1      Thematic Tag2         Thematic Tag3  \
5      alternative energy             biogas  environment friendly   
61                  water        wind energy               pumping   
27  eco-efficient cooking  energy efficiency      renewable energy   

   Thematic Tag4 Thematic Tag5         Lat        Long        image_1 image_2  \
5            NaN           NaN  -12.910466  14.0356608  file: img-324     NaN   
61           NaN           NaN  12.3681873  -1.5270944  file: img-389     NaN   
27           NaN           NaN  -34.606767  -58.377948  file: img-177     NaN   

   image_3 image_4 image_5 image_6 image_7 image_8 image_9  
5      NaN     NaN     NaN     NaN     NaN     NaN     NaN  
61     NaN     NaN     NaN     NaN     NaN     NaN     NaN  
27     NaN     NaN     NaN     NaN     NaN     NaN     NaN

In [120]:
col_index = pd.DataFrame(np.where(energy_data.eq('x'))[1], columns=['col_index'])

In [121]:
col_index.groupby('col_index')['col_index'].count()

col_index
3      82
9     100
10     41
11    127
13     69
14    124
15     64
16     61
Name: col_index, dtype: int64

In [122]:
energy_data.columns[[3,9,10,11,13,14,15,16]]

Index(['Clean Cooking', 'Open Source', 'IP Protected', 'Can Train Others',
       'Prototype', 'Product', 'Market Ready', 'Requires Advanced Order'],
      dtype='object')

In [123]:
energy_data['Clean Cooking'].unique()

array([nan, 'x'], dtype=object)

In [124]:
# # EDITED
# # Replaced x with other values (instead of 1) in below columns instead of 0. This will make it more interpretable by others

# # Columns that will be combined subsequently
# energy_data['Open Source'] = energy_data['Open Source'].replace('x', 'Open Source')
# energy_data['IP Protected'] = energy_data['IP Protected'].replace('x', 'IP Protected')
# energy_data['Prototype'] = energy_data['Prototype'].replace('x', 'Prototype')
# energy_data['Product'] = energy_data['Product'].replace('x', 'Product')

In [87]:
# Other columns that do not need to be combined
# # For these ones, we shall replace x with True and NaN with False
# energy_data[['Clean Cooking', 'Can Train Others', 
#              'Market Ready', 'Requires Advanced Order']] = energy_data[['Clean Cooking', 'Can Train Others', 
#                                                                         'Market Ready', 'Requires Advanced Order']].replace('x', True)
# energy_data[['Clean Cooking', 'Can Train Others', 
#              'Market Ready', 'Requires Advanced Order']] = energy_data[['Clean Cooking', 'Can Train Others', 
#              'Market Ready', 'Requires Advanced Order']].fillna(False)

In [88]:
# energy_data[['Clean Cooking', 'Open Source', 'IP Protected', 'Can Train Others',
#        'Prototype', 'Product', 'Market Ready',
#        'Requires Advanced Order']] = energy_data[['Clean Cooking', 'Open Source', 'IP Protected', 'Can Train Others',
#        'Prototype', 'Product', 'Market Ready',
#        'Requires Advanced Order']].replace('x', True)

In [125]:
energy_data[['Clean Cooking', 'Open Source', 'IP Protected', 'Can Train Others',
       'Prototype', 'Product', 'Market Ready', 'Requires Advanced Order']].sample(10)

Clean Cooking Open Source IP Protected Can Train Others Prototype Product  \
42            NaN         NaN          NaN              NaN       NaN     NaN   
130           NaN           x          NaN                x       NaN       x   
331           NaN         NaN          NaN              NaN       NaN     NaN   
268           NaN           x          NaN                x       NaN       x   
233           NaN         NaN          NaN              NaN       NaN       x   
193           NaN         NaN            x                x       NaN       x   
97              x           x          NaN              NaN       NaN       x   
134           NaN         NaN          NaN              NaN       NaN     NaN   
25            NaN           x          NaN                x         x     NaN   
8             NaN           x          NaN                x         x     NaN   

    Market Ready Requires Advanced Order  
42           NaN                     NaN  
130          NaN                       x  
331          NaN                     NaN  
268          NaN                     NaN  
233            x                     NaN  
193          NaN                       x  
97           NaN                     NaN  
134          NaN                     NaN  
25           NaN                     NaN  
8            NaN                     NaN

In [52]:
# No need replacing nan with false since we will eventually combine some of these columns into long forms
# energy_data[['clean cooking', 'opensource', 'ip_protection', 'replicated',
#        'prototype', 'product', 'market_availability',
#        'advanced_order_applicable']] = energy_data[['clean cooking', 'opensource', 'ip_protection', 'replicated',
#        'prototype', 'product', 'market_availability',
#        'advanced_order_applicable']].applymap(bool)

In [126]:
energy_data.head()

Date    ID          Energy Source Clean Cooking  \
0 2021-05-30 13:48:32  3100                Thermal           NaN   
1 2021-06-28 13:56:30  3177                   Wind           NaN   
2 2021-06-28 13:32:36  3176                  Hydro           NaN   
3 2020-11-12 06:03:43  2261      Renewable general           NaN   
4 2021-07-21 10:12:39  3236  Household application             x   

                           Project Title          Mapper    Contributor  \
0                               Eco Char  AccLab Algeria  Contributor 1   
1                   Wind power generator   AccLab Angola  Contributor 2   
2                       Hybrid Generator   AccLab Angola  Contributor 3   
3  Electric Power Microgeneration System   AccLab Angola  Contributor 4   
4                           Solar Cooker   AccLab Angola  Contributor 5   

                                 Purpose of Solution Link Open Source  \
0  We upcycle wastes such as food organic wastes ...  NaN         NaN   
1  The system to obtain electrical energy through...  NaN         NaN   
2  The production of fuel from water is done thro...  NaN         NaN   
3  The electric power microgeneration system is d...  NaN         NaN   
4   Solar Cooker\n\n\n\t\t\t\n\n\n\t\t\t\n\n\n\t\t\t  NaN         NaN   

  IP Protected Can Train Others Unit Cost in USD Prototype Product  \
0          NaN              NaN              NaN       NaN     NaN   
1          NaN              NaN              NaN       NaN     NaN   
2          NaN              NaN              NaN       NaN     NaN   
3          NaN              NaN              NaN       NaN     NaN   
4          NaN              NaN              NaN       NaN     NaN   

  Market Ready Requires Advanced Order Tech Readiness Level User Feedback  \
0          NaN                     NaN                  NaN           NaN   
1          NaN                     NaN                  NaN           NaN   
2          NaN                     NaN                  NaN           NaN   
3          NaN                     NaN                  NaN           NaN   
4          NaN                     NaN                  NaN           NaN   

  Feedback Link Efficiency Benchmark Potential Bottlenecks  SDG Solution1  \
0           NaN                  NaN                   NaN            1.0   
1           NaN                  NaN                   NaN            1.0   
2           NaN                  NaN                   NaN           10.0   
3           NaN                  NaN                   NaN            1.0   
4           NaN                  NaN                   NaN            7.0   

   SDG Solution2  SDG Solution3  SDG Solution4  SDG Solution5  \
0            6.0           13.0            NaN            NaN   
1            7.0           13.0            NaN            NaN   
2           12.0           13.0            NaN            NaN   
3            7.0            9.0           10.0           11.0   
4           12.0           13.0            NaN            NaN   

           Thematic Tag1          Thematic Tag2          Thematic Tag3  \
0  food waste management                  youth                 coffee   
1     alternative energy           clean energy         climate change   
2                    NaN     alternative energy            electricity   
3           clean energy  community empowerment  rural electrification   
4             innovation     polution reduction           solar energy   

           Thematic Tag4 Thematic Tag5         Lat        Long        image_1  \
0                 olives           NaN  36.7607349   3.0261856            NaN   
1  community empowerment           NaN  -12.833226  13.6010742   file: img-10   
2     polution reduction           NaN  -8.5158356  16.0817871  file: img-121   
3         social justice       welfare  -12.776115   15.687688  file: img-230   
4                    NaN           NaN   -12.86536  13.8098145  file: img-305   

  image_2 image_3 image_4 image_5 image_6 image_

In [127]:
energy_data['User Feedback'].unique()

array([nan,
       "Liter of Light has installed more than 350,000 bottle lights in more than 15 countries and taught green skills to empower grassroots entrepreneurs at every stop. Liter of Light's open source technology has been recognized by the UN and adopted for use in some UNHCR camps. Liter of Light is the proud recipient of the 2016 St. Andrews Prize for the Environment, the 2015 Zayed Future Energy Prize, and a winner of the 2014-2015 World Habitat Award.",
       'So far, over one hundred solar water heaters and their corresponding showers have been installed. In each case, feedback from end users is welcome.',
       'The project has reached over 500 homes directly and over 5,000 homes through its replication in other provinces and countries. In total, it has spread among approximately 2,000 people. In every case, end users can share their experience and give their opinion.',
       ' More than 500 people, including decision makers, experienced an electric bike ride.All of t

In [92]:
energy_data['Clean Cooking'].nunique(dropna=False)

2

In [128]:
energy_data[['Clean Cooking', 'Open Source', 'IP Protected', 'Can Train Others',
       'Prototype', 'Product', 'Market Ready',
       'Requires Advanced Order']].describe()

Clean Cooking Open Source IP Protected Can Train Others Prototype  \
count             82         100           41              127        69   
unique             1           1            1                1         1   
top                x           x            x                x         x   
freq              82         100           41              127        69   

       Product Market Ready Requires Advanced Order  
count      124           64                      61  
unique       1            1                       1  
top          x            x                       x  
freq       124           64                      61

In [129]:
energy_data.isnull().sum()

Date                         2
ID                           0
Energy Source                4
Clean Cooking              277
Project Title                0
Mapper                       0
Contributor                  0
Purpose of Solution         12
Link                       233
Open Source                259
IP Protected               318
Can Train Others           232
Unit Cost in USD           289
Prototype                  290
Product                    235
Market Ready               295
Requires Advanced Order    298
Tech Readiness Level       233
User Feedback              256
Feedback Link              332
Efficiency Benchmark       301
Potential Bottlenecks      306
SDG Solution1               34
SDG Solution2              122
SDG Solution3              181
SDG Solution4              284
SDG Solution5              341
Thematic Tag1               39
Thematic Tag2               77
Thematic Tag3              146
Thematic Tag4              235
Thematic Tag5              314
Lat     

In [95]:
energy_data['Thematic Tag1'].nunique()

71

In [200]:
# Combine Product  and Prototype into one column
m1 = energy_data['Product'] == 'x'
m2 = energy_data['Prototype'] == 'x'
energy_data['Project Stage'] = np.select([m1, m2], ['Product', 'Prototype'], default = 'Unspecified')

# Combine ip protected and open source into one column
m1 = energy_data['Open Source'] == 'x'
m2 = energy_data['IP Protected'] == 'x'
energy_data['Intellectual Property'] = np.select([m1, m2], ['Open Source', 'IP Protected'], default = 'Unspecified')

# return dataframe with only desired tables
energy_data = energy_data.drop(columns = ['Open Source', 'IP Protected', 'Prototype', 'Product'])

In [201]:
energy_data.columns

Index(['Date', 'ID', 'Energy Source', 'Clean Cooking', 'Project Title',
       'Mapper', 'Contributor', 'Purpose of Solution', 'Link',
       'Can Train Others', 'Unit Cost in USD', 'Market Ready',
       'Requires Advanced Order', 'Tech Readiness Level', 'User Feedback',
       'Feedback Link', 'Efficiency Benchmark', 'Potential Bottlenecks',
       'SDG Solution1', 'SDG Solution2', 'SDG Solution3', 'SDG Solution4',
       'SDG Solution5', 'Thematic Tag1', 'Thematic Tag2', 'Thematic Tag3',
       'Thematic Tag4', 'Thematic Tag5', 'Lat', 'Long', 'image_1', 'image_2',
       'image_3', 'image_4', 'image_5', 'image_6', 'image_7', 'image_8',
       'image_9', 'Project Stage', 'Intellectual Property'],
      dtype='object')

In [60]:
# Translate columns to English
# PLEASE RUN THIS LINE OF CODE WHEN YOU HAVE TIME AND GOOD INTERNET CONNECTION.
# energy_data['Project Title'] = energy_data['Project Title'].astype('str').apply(lambda x : ts.google(x, to_language='en', if_ignore_limit_of_length=True))

In [61]:
# energy_data['Project Title'] = energy_data['Project Title'].astype('str').apply(lambda x : ts.google(x, to_language='en', if_ignore_limit_of_length=True))

In [97]:
energy_data[energy_data['Lat'] == 'xxxx']

Date    ID Energy Source  Clean Cooking  \
200 2021-08-24 10:32:22  3458       Thermal          False   

                         Project Title        Mapper      Contributor  \
200  Alttakaddum ou charbon ecologique  AccLab Niger  Contributor 201   

                                   Purpose of Solution  \
200  Cette innovation consiste a transformer les ca...   

                                                  Link Open Source  \
200  https://m.facebook.com/story.php?story_fbid=10...         NaN   

    IP Protected  Can Train Others Unit Cost in USD Prototype Product  \
200          NaN             False              NaN       NaN     NaN   

     Market Ready  Requires Advanced Order Tech Readiness Level User Feedback  \
200         False                    False                  NaN           NaN   

    Feedback Link Efficiency Benchmark Potential Bottlenecks  SDG Solution1  \
200           NaN                  NaN                   NaN            1.0   

     SDG Solution2  SDG Solution3  SDG Solution4  SDG Solution5  \
200            3.0           12.0           13.0            NaN   

          Thematic Tag1 Thematic Tag2               Thematic Tag3  \
200  alternative energy  clean energy  protecting the environment   

           Thematic Tag4     Thematic Tag5   Lat   Long        image_1  \
200  recycle waste plant  waste management  xxxx  xxxxx  file: img-377   

    image_2 image_3 image_4 image_5 image_6 image_7 image_8 image_9  
200     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN

In [98]:
# Replace xxxx and xxxxx with correct cordinate data in both columns respectively. Affected country is Niger republic
energy_data['Lat'] = energy_data['Lat'].replace('xxxx', 17.5968801)
energy_data['Long'] = energy_data['Long'].replace('xxxxx', 8.0828506)

In [99]:
energy_data[energy_data['Lat'] == 'xxxx']

Empty DataFrame
Columns: [Date, ID, Energy Source, Clean Cooking, Project Title, Mapper, Contributor, Purpose of Solution, Link, Open Source, IP Protected, Can Train Others, Unit Cost in USD, Prototype, Product, Market Ready, Requires Advanced Order, Tech Readiness Level, User Feedback, Feedback Link, Efficiency Benchmark, Potential Bottlenecks, SDG Solution1, SDG Solution2, SDG Solution3, SDG Solution4, SDG Solution5, Thematic Tag1, Thematic Tag2, Thematic Tag3, Thematic Tag4, Thematic Tag5, Lat, Long, image_1, image_2, image_3, image_4, image_5, image_6, image_7, image_8, image_9]
Index: []

In [100]:
energy_data[energy_data['Long'] == 'xxxxx']

Empty DataFrame
Columns: [Date, ID, Energy Source, Clean Cooking, Project Title, Mapper, Contributor, Purpose of Solution, Link, Open Source, IP Protected, Can Train Others, Unit Cost in USD, Prototype, Product, Market Ready, Requires Advanced Order, Tech Readiness Level, User Feedback, Feedback Link, Efficiency Benchmark, Potential Bottlenecks, SDG Solution1, SDG Solution2, SDG Solution3, SDG Solution4, SDG Solution5, Thematic Tag1, Thematic Tag2, Thematic Tag3, Thematic Tag4, Thematic Tag5, Lat, Long, image_1, image_2, image_3, image_4, image_5, image_6, image_7, image_8, image_9]
Index: []

In [101]:
energy_data[['Lat', 'Long']] = energy_data[['Lat', 'Long']].astype('float64')

In [102]:
energy_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 359 entries, 0 to 358
Data columns (total 43 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   Date                     357 non-null    datetime64[ns]
 1   ID                       359 non-null    int64         
 2   Energy Source            355 non-null    object        
 3   Clean Cooking            359 non-null    bool          
 4   Project Title            359 non-null    object        
 5   Mapper                   359 non-null    object        
 6   Contributor              359 non-null    object        
 7   Purpose of Solution      347 non-null    object        
 8   Link                     126 non-null    object        
 9   Open Source              100 non-null    object        
 10  IP Protected             41 non-null     object        
 11  Can Train Others         359 non-null    bool          
 12  Unit Cost in USD         70 non-null

In [107]:
# Reshape columns 'Prototype' and 'Product' to long form
id_vars = energy_data.drop(columns = ['Prototype', 'Product']).columns

energy_data = energy_data.melt(id_vars = id_vars, value_name = 'Project Stage')
energy_data = energy_data.drop(columns=['variable'])
energy_data.sample(25)

Date    ID          Energy Source  Clean Cooking  \
384 2022-05-16 19:33:24  4769                Thermal          False   
570 2022-05-11 03:58:27  4670                  Solar          False   
537 2020-10-28 12:44:13  2080  Household application          False   
52  2022-05-16 16:52:02  4751                  Solar          False   
585 2022-05-17 21:45:44  4784  Household application           True   
16  2022-05-03 15:46:38  4589                   Wind          False   
648 2020-09-17 20:41:24   832      Renewable general          False   
85  2022-05-10 14:30:49  4654  Household application           True   
288 2020-09-17 20:53:01   836                Thermal          False   
632 2022-05-29 12:24:45  4925                  Solar          False   
219 2022-05-11 23:58:51  4683  Household application           True   
467 2022-04-07 12:41:30  4385                  Solar          False   
565 2022-05-10 21:28:15  4665                  Solar          False   
322 2022-06-02 06:43:37  4959                  Hydro          False   
124 2022-03-30 17:01:40  4362                Thermal          False   
545 2021-02-24 07:42:38  2769      Renewable general          False   
214 2022-05-11 21:50:46  4678                  Solar          False   
1   2021-06-28 13:56:30  3177                   Wind          False   
487 2021-06-10 13:08:40  3121                Thermal          False   
516 2020-09-15 12:22:57   149      Renewable general          False   
685 2022-06-03 06:25:51  4990  Household application           True   
205 2022-05-10 20:40:33  4661  Household application           True   
258 2022-05-18 09:56:58  4795                 Solar           False   
594 2022-05-24 01:32:54  4825                Thermal          False   
161 2020-09-17 19:36:09   795      Renewable general          False   

                                         Project Title  \
384     Biodigester to obtain gas from livestock waste   
570                Solar-powered water production unit   
537                                   Battery Pack C13   
52                         Mixed dryer (solar and gas)   
585                     Biodigestors for clean cooking   
16      Savonius Wind Turbinebuilt from waste material   
648                                Gastrol green power   
85                                     BY KACHIMATE AT   
288  Biogas from all wastes like food, cow dung and...   
632                 Solar-powered accessories workshop   
219                     Locally designed peanut cooker   
467                               SunCity (solar pump)   
565                            Solar UV drier for kava   
322  Power generation through sewage/slow moving water   
124                         Cow dung beehive briquette   
545                         All-Lights Village Project   
214             First apparel firm to go solar in Fiji   
1                                 Wind power generator   
487                              Urban waste to energy   
516                                              Ezbox   
685               Modified Stove with Kerosene Heating   
205                           Wood gas and solar stove   
258                                   Faith Ildephonse   
594                  Thermoelectric Plant - In Az√∫car   
161                                    Watering System   

                        Mapper      Contributor  \
384           AccLab Argentina   Contributor 26   
570             AccLab Pacific  Contributor 212   
537            AccLab Malaysia  Contributor 179   
52         AccLab Burkina Faso   Contributor 53   
585              AccLab Panama  Contributor 227   
16            AccLab Argentina   Contributor 17   
648              AccLab Uganda  Contributor 290   
85              AccLab Ecuador   Contributor 86   
288              AccLab Uganda  Contributor 289   
632               AccLab Syria  Contributor 274   
219             AccLab Pacific  Contributor 220   
467               AccLab Egypt  Contributor 109   


In [108]:
energy_data[['ID', 'Project Stage']].sample(20)

ID Project Stage
37   4633           NaN
17   4590     Prototype
165  3414           NaN
162  1595           NaN
457  5027       Product
227  4785           NaN
305  2690           NaN
613  2287       Product
57   5112           NaN
75   2849           NaN
414  4944       Product
64   1822           NaN
292   839           NaN
44   5299           NaN
643  5195       Product
357  5092           NaN
1    3177           NaN
197  2674           NaN
559  3458           NaN
54   4932           NaN

In [ ]:
# Reshape columns 'Open Source' and 'IP Protected' to long form
id_vars = energy_data.drop(columns = ['Open Source', 'IP Protected']).columns

energy_data = energy_data.melt(id_vars = id_vars, value_name = 'Intellectual Property')
energy_data = energy_data.drop(columns=['variable'])
energy_data.sample(25)

In [110]:
energy_data.shape

(1436, 41)

In [113]:
energy_data['Project Title']

0                                                Eco Char
1                                    Wind power generator
2                                        Hybrid Generator
3                   Electric Power Microgeneration System
4                                            Solar Cooker
5                 Biogas production through a Biodigester
6       A circuit that provides high quality solar lig...
7       Social plug-ins that provide hot showers: Suma...
8       Hot water for showers and air heating systems:...
9       Solar Electric Mobility Prototype: EcoAndina F...
10      Renewable Network with Storage 24 hours a day:...
11        Solar Heating Savings Kit: EcoAndina Foundation
12                      Solar Kiosk: EcoAndina Foundation
13              UV Solar Streetellites: Single Foundation
14      A container without electricity thatstores veg...
15          An eco-stove built from metal waste: EcoManos
16         Savonius Wind Turbinebuilt from waste material
17      A devi

In [114]:
energy_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1436 entries, 0 to 1435
Data columns (total 41 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   Date                     1428 non-null   datetime64[ns]
 1   ID                       1436 non-null   int64         
 2   Energy Source            1420 non-null   object        
 3   Clean Cooking            1436 non-null   bool          
 4   Project Title            1436 non-null   object        
 5   Mapper                   1436 non-null   object        
 6   Contributor              1436 non-null   object        
 7   Purpose of Solution      1388 non-null   object        
 8   Link                     504 non-null    object        
 9   Can Train Others         1436 non-null   bool          
 10  Unit Cost in USD         280 non-null    object        
 11  Market Ready             1436 non-null   bool          
 12  Requires Advanced Order  1436 non-

In [119]:
# Convert datatypes
convert_dict = {
    'ID' : 'str',
    'Energy Source' : 'category',
    'Clean Cooking' : 'bool',
    'Requires Advanced Order': 'bool',
    'Mapper' : 'category',
    'Can Train Others' : 'bool',
    'Market Ready' : 'bool',
}

energy_data = energy_data.astype(convert_dict)
energy_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1436 entries, 0 to 1435
Data columns (total 41 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   Date                     1428 non-null   datetime64[ns]
 1   ID                       1436 non-null   object        
 2   Energy Source            1420 non-null   category      
 3   Clean Cooking            1436 non-null   bool          
 4   Project Title            1436 non-null   object        
 5   Mapper                   1436 non-null   category      
 6   Contributor              1436 non-null   object        
 7   Purpose of Solution      1388 non-null   object        
 8   Link                     504 non-null    object        
 9   Can Train Others         1436 non-null   bool          
 10  Unit Cost in USD         280 non-null    object        
 11  Market Ready             1436 non-null   bool          
 12  Requires Advanced Order  1436 non-

### QUESTIONS
1. Where are the solutions coming from? What is their distribution per country & per region?

2. What type of energy source is more prevalent, what is less? Are there differences per region, and why?

3. What are global commonalities across solutions; what are typical applications & use cases for solutions? Are there patterns that emerge when looking at the distribution per country & per region? 

4. What overall challenges are the solutions addressing or contributing to overcome?

5. Which Sustainable Development Goals are the solutions advancing in particular, and how?

6. Looking at the use case of clean cooking solutions, what is their prevalence, distribution, and source of energy?
                

If applicable / information is available*:

- How are the solutions solving a specific community problem? What are the solutions addressing or contribute to solving? (if applicable)

- How can we display the more quantitative information (ratio of IP vs DIY solutions; Prototype vs Product) in an appealing way that signifies the availability of solutions in country? (if applicable)

In [ ]:
# Where are the solutions coming from? What is their distribution per country & per region?
